### ISSUE TITLE
Multi-token decode produces garbage on Gemma-4 PLE models

### ISSUE THREAD

**Issue by deem0n** [chrishayuk/larql#98]
# Multi-token decode produces garbage on Gemma-4 PLE models — KV-cached generate path desyncs PLE per-token inputs

  ## TL;DR

  After #49 (PR #91) restored PLE sidecar capture on `--quant none` extracts, `larql lql INFER` produces coherent high-confidence
  output on `gemma-4-E4B-it`. But the KV-cached `generate_cached` path used by `larql run` and `larql chat` still produces garbage on
  multi-token decode — the **first** decode token after prefill is correct; every subsequent token degrades into Gemma-4's "broken
  PLE" failure signature (CJK fragments, code identifiers, mid-word fragments).

  This is a separate root cause from #49. The writer / loader / forward-path PLE wiring are all correct; the bug is in how the decode
  loop extends `per_layer_inputs` across newly generated tokens.

  ## Reproducer

  ```bash
  # Extract using PR #91 (gemma4-ple-noquant-fix) — PLE sidecars all land:
  larql extract /path/to/gemma-4-E4B-it -o e4b.vindex --summary-features-per-expert 64

  # LQL INFER one-shot — works correctly (validates #49 fix):
  larql lql 'USE "e4b.vindex"; INFER "The capital of France is" TOP 5;'
  #   1.  France              (99.68%)
  #   2.  French              (0.07%)
  #   3.  is                  (0.05%)
  #   4.  The                 (0.03%)
  #   5.  the                 (0.02%)

  # Single-token decode via `larql run` — also correct:
  larql run e4b.vindex "The capital of France is" --max-tokens 1
  # France

  # Multi-token decode — garbage from token 2 onward:
  larql run e4b.vindex "The capital of France is" --max-tokens 5
  # ッケッケTobchal的存在
  ```

  `ッケ` was one of the pre-fix garbage tokens from #49. Seeing it re-appear specifically on the second decode token is the smoking
  gun: PLE *was* applied during prefill (token 1 is "France", consistent with INFER), then *isn't* during the next decode step.

  ## Two paths, two outcomes

  | Path | First token | Subsequent tokens |
  |---|---|---|
  | `larql lql INFER` (`predict_q4k_hidden_with_ffn` → walk FFN) | Correct (`France`) | N/A (single-shot) |
  | `larql run --max-tokens 1` (`generate_cached` → full FFN, prefill only) | Correct (`France`) | N/A |
  | `larql run --max-tokens N>1` (`generate_cached`, prefill + decode loop) | Correct (`France`) | Garbage (`ッケッケTobchal的存在`) |

  Prefill is fine — three independent paths through prefill all converge on the right token. Decode is where it breaks.

  ## Hypothesis

  `forward/ple.rs::precompute_per_layer_inputs` takes a full `main_embeds: [seq, hidden]` and `token_ids: [seq]`, and returns one
  `[seq, ple_dim]` per layer. The KV-cached decode loop (likely in `crates/larql-inference/src/forward/kv_generate.rs` or
  `layer_graph::generate`) appears to:

  1. Compute `per_layer_inputs` for the prompt prefix.
  2. Append the newly generated token's embed to `main_embeds`.
  3. Re-run `apply_per_layer_embedding(per_layer_input)` against the *same prefix-sized* `per_layer_input`, indexing position `seq_len
   - 1` which is **out of bounds** of the precomputed inputs — or worse, indexing into stale data from another decode step.

  Symptoms consistent with that:
  - Token 1 (`seq_len == prompt_len`): `per_layer_input` covers position 0..prompt_len-1; last position lookup hits prompt_len-1 —
  valid. Output coherent.
  - Token 2+ (`seq_len > prompt_len`): lookup at position prompt_len reads past the precomputed slab → garbage; or reads a stale row
  that no longer corresponds to the actual decode token.

  Code to inspect:
  - `crates/larql-inference/src/forward/ple.rs::precompute_per_layer_inputs` — verify the return shape contract.
  - `crates/larql-inference/src/forward/kv_generate.rs` (or wherever `generate_cached` lives) — check whether `per_layer_inputs` is
  recomputed / appended-to per decode step, vs. computed once over the prompt and reused.
  - Walk path (`predict_q4k_hidden_with_ffn`) for the contrast: it likely re-computes PLE for the single-position decode each call,
  which is why one-shot INFER works.

  ## Why this isn't covered by #49's tests

  The PR #91 regression tests assert PLE tensors land in `weight_manifest.json` and hydrate into `weights.tensors` /
  `weights.vectors`. They don't drive a multi-token decode against a real model — the synthetic fixture is sized for byte-layout
  coverage, not numerical correctness over a generation loop.

  ## Out of scope (but related)

  #49 also called out three adjacent issues. This bug is distinct from all three:

  - **A** (`WalkFfn::forward_moe_full_layer` missing) — applies to hybrid-MoE Gemma-4 (e.g. `gemma-4-26B-A4B-it`). E4B is dense,
  doesn't route through MoE. The garbage here is on a dense PLE model.
  - **B** (`predict_with_ffn_attention` shared-KV) — about `EXPLAIN INFER WITH ATTENTION`. Doesn't touch `generate_cached`.
  - **C** (LQL INFER chat-wrap) — explains why "Paris" isn't top-1 on the working path. Doesn't affect whether subsequent tokens are
  coherent vs. garbage.

  ## Suggested fix scope

  1. Make the decode loop responsible for extending `per_layer_inputs` per token, OR
  2. Compute the per-token PLE input inline in the decode step from `(weights, new_token_id)` rather than indexing a precomputed slab,
   OR
  3. Recompute the whole `per_layer_inputs` over `[prompt..current]` each decode step (slower but simplest; the cost is one matmul
  against the prompt-len embed table — small).

  Whichever fix lands should add a multi-token decode test against the synthetic Gemma-4 PLE fixture (`write_gemma4_ple_fixture` in
  `crates/larql-vindex/tests/test_vindex.rs`) — generate ≥3 tokens and assert all tokens decode to non-empty UTF-8 strings whose
  embedding-cosine to a known anchor token doesn't collapse to noise. The synthetic fixture's data is uniform so it won't reproduce
  the exact `ッケ` mode, but the *kind* of failure (logits saturating on a single weird id, or NaN propagating) will be caught.

  ## Validated on

  - Branch: `chrishayuk/larql:main` + PR #91 (commit 7d50997).
  - Model: `google/gemma-4-E4B-it`, bf16 safetensors from HuggingFace.
  - Extract: `--quant none --summary-features-per-expert 64 --level inference`.
  - macOS 14, 32 GB RAM, CPU backend (no Metal).
  - larql build: `cargo build --release -p larql-cli --no-default-features`.
